### PR TITLE
Multi server

### PR DESCRIPTION
**Pull request by ak08743**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -7,6 +7,13 @@ Optune servo driver for Prometheus
 2. The default Prometheus endpoint is `http://prometheus:9090/metrics`; you must
    set the environment variable `PROMETHEUS_ENDPOINT` in the servo's container
    if you want to use a different endpoint.
+3. Alternatively, Prometheus endpoint may be specified in `prometheus_endpoint` key
+   of driver configuration. If this key exists and is not empty, value specified will
+   override `PROMETHEUS_ENDPOINT`.
```
Comment (Johnius) [inline README.md:12]: Personally I would prefer to have the environment variable to take precedence.

File: measure [hunk 2]
```diff
@@ -9,86 +9,112 @@ import yaml
...
+        #   defined in section matching driver name in that case.
+        driver_key = "prom"
+        if os.environ.get('PROMETHEUS_USE_DRIVER_NAME', False):
+            driver_key = os.path.basename(__file__)
         try:
-            config = yaml.safe_load(open(path))['prom']
-            assert config and config.get('metrics'), 'No metrics has been provided at path "prom.metrics" ' \
-                                                     'in file located at {}'.format(path)
+            config = yaml.safe_load(open(path))[driver_key]
+            assert config and config.get("metrics"), 'No metrics has been provided at path "{}.metrics" in file located at {}"'.format(driver_key, path)
```
Comment (Johnius) [inline measure:36]: FYI, we can now leverage format `f'{config.get('metric')} is not found'`

File: measure [hunk 3]
```diff
@@ -9,86 +9,112 @@ import yaml
...
+        metrics = config["metrics"]
+        assert isinstance(metrics, dict), "The `metrics` attribute in `{}` section must be a dict".format(driver_key)
+        for m, mdesc in config["metrics"].items():
+            assert isinstance(m, str), "Metric name `{}` must be a string, found {}".format(m, type(m).__name__)
             if not isinstance(mdesc, dict):
-                raise Exception('Metric `{}` data in file {} must be a dict, found {}'.
-                    format(m, path, type(mdesc).__name__))
-            bad_keys = mdesc.keys() - {'query', 'unit'}
+                raise Exception(
+                    "Metric `{}` data in file {} must be a dict, found {}".format(
```
Comment (Johnius) [inline measure:63]: I would name `data` - `settings`

File: measure [hunk 4]
```diff
@@ -9,86 +9,112 @@ import yaml
...
+            return config
+
+        def get_prom_metrics(self):
+            """
+            Get metrics known to Prometheus's endpoint
+            (Currently not used)
+            """
+            base_url = self.get_config().get("prometheus_endpoint", PROM_BASE_URL).rstrip("/")
+
+            url = "%s%slabel/__name__/values" % (base_url, PROM_API_PATH)
```
Comment (Johnius) [inline measure:94]: I believe we decided to ditch `%` approach company-wide

File: README.md (hunk 1 above)
Comment (ak08743) [reply README.md:12]: Environment variable is not applicable when multiple drivers are used under MAGG, as the value would apply to all drivers, defying the purpose. Defining the precedence as 'env var unless you do multiple drivers, then config value' leads to inconsistent behaviour that, to my mind, is worse.

File: measure (hunk 2 above)
Comment (ak08743) [reply measure:36]: Right, but this comes from original driver - I'm not modifying that part.

File: measure (hunk 4 above)
Comment (ak08743) [reply measure:94]: This is a part of the original driver, not changing it.

File: measure (hunk 3 above)
Comment (ak08743) [reply measure:63]: Can you please explain what do you mean ?

File: README.md [hunk 5]
```diff
@@ -7,6 +7,13 @@ Optune servo driver for Prometheus
...
+4. If environment variable `PROMETHEUS_USE_DRIVER_NAME` is set to True, then
```
Comment (pnickolov) [inline README.md]: "1" instead of True?

Comment (ak08743) [reply README.md]: Fixed

**Pull request merged**